homeassistant/components/tplink/config_flow.py

```diff
@@ -137,16 +320,111 @@ def _async_create_entry_from_device(self, device: SmartDevice) -> FlowResult:
             title=f"{device.alias} {device.model}",
             data={
                 CONF_HOST: device.host,
+                CONF_ALIAS: device.alias,
+                CONF_MODEL: device.model,
+                CONF_DEVICE_CONFIG: device.config.to_dict(
+                    credentials_hash=device.credentials_hash,
+                    exclude_credentials=True,
+                ),
             },
         )
 
+    async def _async_try_discover_and_update(
+        self,
+        host: str,
+        credentials: Optional[Credentials],
+        raise_on_progress: bool,
+    ) -> SmartDevice:
+        """Try to connect."""
+        try:
+            self._discovered_device = await Discover.discover_single(
+                host, credentials=credentials
+            )
+        except TimeoutException:
+            # Try connect() to legacy devices if discovery fails
+            self._discovered_device = await SmartDevice.connect(
+                config=DeviceConfig(host)
+            )
+        else:
+            if self._discovered_device.config.uses_http:
+                self._discovered_device.config.http_client = (
+                    create_async_tplink_clientsession(self.hass)
+                )
+            await self._discovered_device.update()
+        await self.async_set_unique_id(
+            dr.format_mac(self._discovered_device.mac),
+            raise_on_progress=raise_on_progress,
+        )
+        return self._discovered_device
+
     async def _async_try_connect(
-        self, host: str, raise_on_progress: bool = True
+        self,
+        discovered_device: SmartDevice,
+        credentials: Optional[Credentials],
     ) -> SmartDevice:
         """Try to connect."""
-        self._async_abort_entries_match({CONF_HOST: host})
-        device: SmartDevice = await Discover.discover_single(host)
+        self._async_abort_entries_match({CONF_HOST: discovered_device.host})
+
+        config = discovered_device.config
+        if credentials:
+            config.credentials = credentials
+        config.timeout = CONNECT_TIMEOUT
+        if config.uses_http:
+            config.http_client = create_async_tplink_clientsession(self.hass)
+
+        self._discovered_device = await SmartDevice.connect(config=config)
         await self.async_set_unique_id(
-            dr.format_mac(device.mac), raise_on_progress=raise_on_progress
+            dr.format_mac(self._discovered_device.mac),
+            raise_on_progress=False,
+        )
+        return self._discovered_device
+
+    async def async_step_reauth(self, entry_data: Mapping[str, Any]) -> FlowResult:
+        """Handle reauth upon an API authentication error."""
```

Comment:
Docstring could be more informative.

Reply:
Updated to `Start the reauthentication flow if the device needs updated credentials.`